src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs

```diff
@@ -25,6 +26,8 @@ internal class AnalyzerExecutor
         private readonly Action<Exception, DiagnosticAnalyzer, Diagnostic> _onAnalyzerException;
         private readonly AnalyzerManager _analyzerManager;
         private readonly Func<DiagnosticAnalyzer, bool> _isCompilerAnalyzer;
+        private readonly ImmutableDictionary<DiagnosticAnalyzer, object> _singleThreadedAnalyzerToGateMap;
```

Comment:
Analyzers can call back into the compilation, force completing symbols, generating compilation events or querying diagnostics. But none of these can ever get back to invoking the driver or the executor. All the event processing and subsequent analyzer execution will happen on background threads processing compilation events.

Only case is probably if the analyzer casts the handed compilation to CompilationWithAnalyzers and waits on GetAnalyzerDiagnostics(), but that would anyways block forever.